File: apps/ios/Janark/Core/Outcome.swift
import Foundation

enum Outcome<T> {
    case ok(T)
    case err(String, unauthorized: Bool = false, status: Int? = nil)

    var value: T? {
        if case .ok(let v) = self { return v }
        return nil
    }

    var errorMessage: String? {
        if case .err(let message, _, _) = self { return message }
        return nil
    }

    var isUnauthorized: Bool {
        if case .err(_, let unauthorized, _) = self { return unauthorized }
        return false
    }

    func getOrNull() -> T? { value }

    func map<R>(_ transform: (T) -> R) -> Outcome<R> {
        switch self {
        case .ok(let v): return .ok(transform(v))
        case .err(let message, let unauthorized, let status):
            return .err(message, unauthorized: unauthorized, status: status)
        }
    }
}

File: apps/android/app/src/main/java/org/janark/app/core/Outcome.kt
package org.janark.app.core

sealed class Outcome<out T> {
    data class Ok<T>(val value: T) : Outcome<T>()
    data class Err(val message: String, val unauthorized: Boolean = false, val status: Int? = null) :
        Outcome<Nothing>()
}

inline fun <T> Outcome<T>.getOrNull(): T? = (this as? Outcome.Ok)?.value

inline fun <T, R> Outcome<T>.map(transform: (T) -> R): Outcome<R> = when (this) {
    is Outcome.Ok -> Outcome.Ok(transform(value))
    is Outcome.Err -> this
}
